### PR TITLE
Icons in MainWidget

### PR DESCRIPTION
**Pull request by StasDash**
- Title is now separate widget
- Some refactoring
- Some documentation
- Added icons
- Maybe fixed some scaling issues

Due to some limitations icons use regular pointers as (icon) override, so nullptr will be used as override remover.

### CODE REVIEW

File: Client/Client.Qt/MainWidget.cpp [hunk 1]
```diff
@@ -233,7 +233,8 @@ void MainWidget::mouseReleaseEvent(QMouseEvent* event)
 
 MainWidget::MainWidget(QWidget* parent) : QWidget(parent)
 {
-    Style::setDpiScale(logicalDpiX() * 100 / 96);
+    Style::setDevicePixelRatio(devicePixelRatioF());
+    Style::setDpiScale(logicalDpiX() * devicePixelRatioF() * 100 / 96);
```
Comment (fisyamaloy) [inline Client/Client.Qt/MainWidget.cpp:237]: What are the magic numbers 100 and 96?

Comment (StasDash) [reply Client/Client.Qt/MainWidget.cpp:237]: Default dpi is 96(100%)

File: Client/Client.Qt/registration.cpp [hunk 2]
```diff
@@ -1,37 +1,41 @@
 #include "registration.hpp"
-#include "Utility/UserDataValidation.hpp"
 
 #include <QtEvents>
+#include <iostream>
```
Comment (Svidunovich) [inline Client/Client.Qt/registration.cpp:4]: Do we need iostream header in this place?

File: Client/Client.Qt/main.cpp [hunk 3]
```diff
@@ -1,4 +1,7 @@
+#include <QDebug>
```
Comment (Svidunovich) [inline Client/Client.Qt/main.cpp:1]: Do we need qdebug header in this place?

File: Client/Client.Qt/Widgets/ScrollArea.cpp [hunk 4]
```diff
@@ -399,8 +382,7 @@ void ScrollArea::moveEvent(QMoveEvent* e)
 
 void ScrollArea::keyPressEvent(QKeyEvent* e)
 {
-    if ((e->key() == Qt::Key_Up || e->key() == Qt::Key_Down) &&
-        e->modifiers().testFlag(Qt::AltModifier))
+    if ((e->key() == Qt::Key_Up || e->key() == Qt::Key_Down) && e->modifiers().testFlag(Qt::AltModifier))
     {
         e->ignore();
     }
```
Comment (Svidunovich) [inline Client/Client.Qt/Widgets/ScrollArea.cpp:388]: image

File: Client/Client.Qt/Widgets/MenuItem.cpp [hunk 5]
```diff
@@ -1,32 +1,31 @@
...
+void ItemBase::paintEvent(QPaintEvent* paintEvent)
 {
     QPainter painter(this);
     painter.setPen(Qt::NoPen);
     QColor color = QColor(0x32, 0x32, 0x32);
-    if (hit)
+    if (isOver())
     {
         color = color.lighter(175);
     }
```
Comment (Svidunovich) [inline Client/Client.Qt/Widgets/MenuItem.cpp:31]: image

File: Client/Client.Qt/Widgets/Menu.cpp [hunk 6]
```diff
@@ -2,32 +2,33 @@
...
+    resize(Style::valueDPIScale(250), _items.back()->y() + _items.back()->height());
 }
 
 void Menu::hideEvent(QHideEvent* hideEvent)
 {
-    if (parent())
+    if (parentWidget())
     {
         parentWidget()->hide();
     }
```
Comment (Svidunovich) [inline Client/Client.Qt/Widgets/Menu.cpp:34]: image

File: Client/Client.Qt/Widgets/CaptionButton.cpp [hunk 7]
```diff
@@ -3,95 +3,64 @@
...
+    }
     QWidget::paintEvent(event);
 }
+
+bool CaptionButton::setIcon(const Style::icon* icon)
+{
+    if (icon != nullptr)
+    {
+        if (icon->size().width() != icon->size().height()) return false;
+    }
```
Comment (Svidunovich) [inline Client/Client.Qt/Widgets/CaptionButton.cpp:63]: image

File: Client/Client.Qt/Widgets/CaptionButton.cpp [hunk 8]
```diff
@@ -3,95 +3,64 @@
...
+    const Style::icon& icon    = _iconOverride ? *_iconOverride : _icon;
 
-    p.drawPixmap((width() - pixmap->width()) / 2, (height() - pixmap->height()) / 2, *pixmap);
+    const int horside = (width() - maxSide) / 2;
+    const int verside = (height() - maxSide) / 2;
+
+    if (icon.size().width() <= maxSide)
+    {
+        painter.drawPixmap((width() - icon->pixmap()->width()) / 2, (height() - icon->pixmap()->height()) / 2, *icon->pixmap());
+    }
```
Comment (Svidunovich) [inline Client/Client.Qt/Widgets/CaptionButton.cpp:49]: image

File: Client/Client.Qt/Style/StyleIcon.cpp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        _pixmap = QPixmap::fromImage(_image);
+        _size   = _image.size();
+    }
+}
+
+QImage IconData::instance() const
+{
+    if (!_image.isNull())
+    {
```
Comment (Svidunovich) [inline Client/Client.Qt/Style/StyleIcon.cpp:30]: image

**Review by Svidunovich — CHANGES_REQUESTED** [review]

File: Client/Client.Qt/Widgets/ScrollArea.cpp (hunk 4 above)
Comment (StasDash) [reply Client/Client.Qt/Widgets/ScrollArea.cpp:388]: Не обязательны не равно не надо

File: Client/Client.Qt/registration.cpp (hunk 2 above)
Comment (StasDash) [reply Client/Client.Qt/registration.cpp:4]: There is below used as debug

**Review by Svidunovich — APPROVED** [review]

**Review by fisyamaloy — APPROVED** [review]

**Pull request merged**